api/explorer/api/ingest.py

```diff
@@ -101,89 +102,99 @@ def _edge_from_build_input_type(build_input_type: model.BuildInputType) -> graph
-
-    Raises:
-        Exception: when the output path of an edge's vertex does not correspond to
-    any derivation in the NixGraph
+            list[graph.Package],
+            list[tuple[graph.Package, graph.Edge, graph.Package]]:
+        ]
+        The packages and edges created while processing nix_package. In the edges,
+        a tuple (A, edge, B) means A -edge-> B.
     """
-    # Map from output path to package node
-    packages: dict[str, graph.Package] = {}
-
-    # A dictionary that maps pairs of package output paths to edge objects.
-    edge_refs = {}
-
-    # A list of tuples containing package, edge, and dependency package objects.
-    edges = []
-
-    # Iterate over each package in the Nix graph
-    for p in nix_graph.packages:
-        parsed = safe_parse_package(p)
-        for ps in parsed:
-            packages[ps.outputPath] = ps
-            # Populate edge_refs dictionary with package output path pairs
-            # and their corresponding edge objects
-            edge_refs |= {
-                (
-                    ps.outputPath,
-                    build_input.output_path,
-                ): _edge_from_build_input_type(build_input.build_input_type)
-                for build_input in p.build_inputs
-            }
-    missing = []
-    for paths, edge in edge_refs.items():
-        package_output_path, dependency_output_path = paths
-        if package_output_path not in packages:
-            raise Exception(
-                "Package metadata could not be found for the output path "
-                "corresponding to the edge's input vertex. "
-                "This should be impossible! "
-                f"Output path: {package_output_path} "
-            )
-        if dependency_output_path not in packages:
-            missing.append(dependency_output_path)
+    edges: list[tuple[graph.Package, graph.Edge, graph.Package]] = []
+
+    parsed_pkgs: list[graph.Package] = safe_parse_package(nix_package)
+    for pkg in parsed_pkgs:
+        ctxt.packages[pkg.outputPath] = pkg
+
+        # Go over the build inputs of the package being processed, and
+        # record edges to these build inputs.
+        for build_input in nix_package.build_inputs:
+            edge: graph.Edge = _edge_from_build_input_type(build_input.build_input_type)
+            dest_output_path: str = build_input.output_path
+            dest: Optional[graph.Package] = ctxt.packages.get(dest_output_path, None)
+            if dest:
+                # Record edge right away
+                edges.append((pkg, edge, dest))
+            else:
+                # Register for being recorded when destination will be processed
+                if dest_output_path in ctxt.pending_edges:
+                    ctxt.pending_edges[dest_output_path].append((pkg, edge))
+                else:
+                    ctxt.pending_edges[dest_output_path] = [(pkg, edge)]
+
+    # For every new package, add pending edges to this package
```

Comment:
What is a "pending" edge?

Reply:
It's an edge for which we have the `graph.Package` source available, but not yet the destination `graph.Package`. This happens, because we process packages one by one now, so we can't have them all created beforehand.

Reply:
I've written more precise comments here: https://github.com/tweag/nixpkgs-graph-explorer/pull/110#issuecomment-1569908404

Let's do it in another PR